ports/mimxrt/machine_rtc.c

```diff
@@ -156,11 +267,158 @@ STATIC mp_obj_t machine_rtc_calibration(mp_obj_t self_in, mp_obj_t cal_in) {
+        machine_rtc_alarm_off(true);
+        if (SNVS_LP_SRTC_SetAlarm(SNVS, &srtc_date) != kStatus_Success) {
+            mp_raise_ValueError(NULL);
+        }
+        machine_rtc_alarm_on();
+    } else { // then it must be an integer
+        seconds = (args[1].u_obj == mp_const_none) ? 0 : (mp_obj_get_int(args[1].u_obj) / 1000);
+        machine_rtc_alarm_helper(seconds, repeat);
+    }
+
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(machine_rtc_alarm_obj, 1, machine_rtc_alarm);
+
+STATIC mp_obj_t machine_rtc_alarm_left(size_t n_args, const mp_obj_t *args) {
+    // only alarm id 0 is available
+    if (n_args > 1 && mp_obj_get_int(args[1]) != 0) {
+        mp_raise_OSError(MP_ENODEV);
+    }
+    uint32_t seconds = machine_rtc_get_seconds();
+    uint32_t alarmSeconds = SNVS->LPTAR;
+    return mp_obj_new_int((alarmSeconds >= seconds) ? ((alarmSeconds - seconds) * 1000) : 0);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(machine_rtc_alarm_left_obj, 1, 2, machine_rtc_alarm_left);
+
+STATIC mp_obj_t machine_rtc_alarm_cancel(size_t n_args, const mp_obj_t *args) {
+    // only alarm id 0 is available
+    if (n_args > 1 && mp_obj_get_int(args[1]) != 0) {
+        mp_raise_OSError(MP_ENODEV);
+    }
+    machine_rtc_alarm_off(true);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(machine_rtc_alarm_cancel_obj, 1, 2, machine_rtc_alarm_cancel);
+
+STATIC mp_obj_t machine_rtc_irq(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_trigger, ARG_handler, ARG_wake };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_trigger, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_handler, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        { MP_QSTR_wake, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // check we want alarm0
+    if (args[ARG_trigger].u_int != 0) {
+        mp_raise_OSError(MP_ENODEV);
+    }
+
+    machine_rtc_irq_obj_t *irq = MP_STATE_PORT(machine_rtc_irq_object);
+
+    // Allocate the IRQ object if it doesn't already exist.
+    if (irq == NULL) {
+        irq = m_new_obj(machine_rtc_irq_obj_t);
+        irq->base.base.type = &mp_irq_type;
+        irq->base.methods = (mp_irq_methods_t *)&machine_rtc_irq_methods;
+        irq->base.parent = MP_OBJ_FROM_PTR(pos_args[0]);
+        irq->base.handler = mp_const_none;
+        irq->base.ishard = true;
+        MP_STATE_PORT(machine_rtc_irq_object) = irq;
+    }
+
+    machine_rtc_alarm_off(false);
+
+    irq->base.handler = args[ARG_handler].u_obj;
+
+    if (SNVS->LPTAR) {
+        machine_rtc_alarm_on();
+    }
+
+    return MP_OBJ_FROM_PTR(irq);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(machine_rtc_irq_obj, 1, machine_rtc_irq);
+
+STATIC mp_obj_t machine_rtc_wakeup(size_t n_args, const mp_obj_t *args) {
```

Comment:
I'm not sure why this method is added, it's not part of the `machine.RTC` documentation.

Does it add any functionality that's not already available by the other alarm methods?  Can it be removed?

Reply:
CC @kwagyeman

Reply:
I updated this PR to fix all comments. Not sure about `machine_rtc_wakeup`, @kwagyeman should know more.

Reply:
There is no consistent machine RTC API. If you look at the different boards they all have slightly different APIs that are actually implemented. So, I tried to implement a super-set of all of them.

Reply:
I understand that some ports don't follow the docs, but we need to make sure that new ports do follow them (and update the docs if needed).

For this particular method, it looks like it's just duplicating functionality of the `.irq()` method.  In that case this method can be removed and there's no loss in functionality.

Reply:
I think `wakeup()` combines the functionality of `alarm()` and `irq()` (i.e configures RTC alarm after `ms`), I don't know if it make sense to have it, but it is for sure more convenient to setup the alarm and callback by calling just one function. Note the same function exists in `stm32` and `ra`, but none of the other documented ones, so maybe it can be added to the RTC API so we can have a common API to use ?

Reply:
The wakeup method seems to have this signature:
```
rtc.wakeup(ms, handler=None)
```
and is equivalent to:
```
rtc.irq(handler=handler)
rtc.alarm(time=ms, repeat=True)
```
It's pretty easy to use the second form, there's really no need for `rtc.wakeup()`.

And in the second form you don't need to register the irq handler once it's been set.  You can simply call `rtc.alarm()` each time you need to reset it.  Also probably you don't want it to repeat, since it's intended to wake-up the device from sleep mode.

Note that `stm32` has `wakeup()` because it doesn't have `irq()` or `alarm()`.  The intention was for the latter pair of functions to replace the former, and be more general.

Reply:
> It's pretty easy to use the second form, there's really no need for rtc.wakeup().

That's true, however until `irq` and `alarm` are added to `stm32` and `ra`, scripts will have to check if `wakeup` exists or not, if not use `irq` + `alarm`, to run on all 3 ports using `machine`, which is not that bad, but defeats the purpose of using `machine`.

Reply:
I removed `wakeup` from this PR, but we will keep it in our fork for a while until all ports we use have a common `machine.RTC` API.